File: C++/check-digitorial-permutation.cpp
// Time:  O(logn + 10)
// Space: O(10)

// freq table
vector<int> FACT(2, 1);
const auto& factorial = [](int n) {
    while (size(FACT) <= n) {
        FACT.emplace_back(FACT.back() * size(FACT));
    }
    return FACT[n];
};

class Solution {
public:
    bool isDigitorialPermutation(int n) {
        const auto& f = [](int x) {
            vector<int> cnt(10);
            for (; x; x /= 10) {
                ++cnt[x % 10];
            }
            return cnt;
        };

        int total = 0;
        for (int x = n; x; x /= 10) {
            total += factorial(x % 10);
        }
        return f(total) == f(n);
    }
};

File: Python/check-digitorial-permutation.py
# Time:  O(logn + 10)
# Space: O(10)

# freq table
FACT = [1]*2
def factorial(n):
    while len(FACT) <= n:
        FACT.append(FACT[-1]*len(FACT))
    return FACT[n]

class Solution(object):
    def isDigitorialPermutation(self, n):
        """
        :type n: int
        :rtype: bool
        """
        def f(x):
            cnt = [0]*10
            while x:
                x, r = divmod(x, 10)
                cnt[r] += 1
            return cnt

        total = 0
        x = n
        while x:
            x, r = divmod(x, 10)
            total += factorial(r)
        return f(total) == f(n)
